File: uniquenumberofoccurrences.js
//Objective is to see whether a 1-D array has unique occurrences for each
//unique element in the array.

let nums = [1,2,2,1,1,3]

//O(n) solution that uses a hashmap to get the frequencies of each element
//and a hashset to make sure no frequency happens twice.

let map = {}

for (let num of nums) {
    if (map[num] == undefined) {
        map[num] = 1
    } else {
        map[num]++
    }
}

let set = new Set()

for (let frequency of Object.values(map)) {
    if (!set.has(frequency)) {
        set.add(frequency)
    } else {
        return false
    }
}

return true

File: uniquenumberofoccurrences.java
//Java Solution

class Solution {
    public boolean uniqueOccurrences(int[] arr) {
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int val : arr) {
            map.put(val, map.getOrDefault(val, 0) + 1);
        }
        
        Set<Integer> set = new HashSet<>();
        
        for (int val : map.values()) {
            if (!set.contains(val)) {
                set.add(val);
            } else {
                return false;
            }
        }
        
        return true;
    }
}